core/src/connection_reuse.rs

```diff
@@ -267,39 +267,42 @@ where
         // Check for any incoming connection on the listening socket.
         // Note that since `self.listener` is a `Fuse`, it's not a problem to continue polling even
         // after it is finished or after it error'ed.
-        match self.listener.poll() {
-            Ok(Async::Ready(Some(upgrade))) => {
-                self.current_upgrades.push(upgrade);
-            }
-            Ok(Async::NotReady) => {}
-            Ok(Async::Ready(None)) => {
-                debug!("listener has been closed");
-                if self.connections.is_empty() && self.current_upgrades.is_empty() {
-                    return Ok(Async::Ready(None));
+        loop {
```

Comment:
Can this be extracted to a trait method like `poll_to_end(on_some_fn, on_none_fn) -> Result<(), WhateverErr>`? Since we seem to need this all over the codebase it might be useful to have it abstracted rather than having an error-prone pattern.

Reply:
IMO the `poll_to_end` version would be more difficult to read and more error-prone.

Reply:
I can see how it would be less readable but I don't know if it would be more error-prone

Reply:
FWIW I also prefer the loops in here.